File: lib/client/go/img_intercept_storelog.go
// main.go
package main
import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"cloud.google.com/go/storage"
	"github.com/google/martian/v3"
	"github.com/google/martian/v3/fifo"
	"github.com/google/martian/v3/httpspec"
	"github.com/google/martian/v3/mitm"
	"github.com/google/martian/v3/proxyutil"
	"google.golang.org/api/option"

	"github.com/dongkyu20/img2webp_proxy/lib/client/go/file_exists"
)

// 전역 설정
const (
	// 사용자 설정
	CDNBaseURL           = "https://storage.cloud.google.com/cdn.ecarbon.kr" // 실제 CDN 주소로 변경
	BucketName           = "cdn.ecarbon.kr"                                  // GCS 버킷 이름
	UpdateInterval       = 3600                                              // 파일 목록 업데이트 주기 (초) - 1시간

	// 원격 로깅 설정
	EnableRemoteLogging    = true                           // 원격 로깅 활성화 여부
	RemoteLogServerURL     = "http://211.253.31.134:5000/log" // PC B의 실제 IP 주소로 변경
	RemoteLogTimeoutSeconds = 1                            // 원격 로깅 요청 타임아웃
)

var (
	// 이미지 확장자를 확인하는 정규식
	originalImageExtRegex = regexp.MustCompile(`\.(png|jpe?g)(\?.*)?$|[_=](png|jpe?g)($|\?|&)|atchFileId=.*_(png|jpe?g)($|\?|&)`)
	
	// CDN 파일 목록을 저장하는 텍스트 파일
	cdnFileListPath = "cdn_file_list.txt"
	
	// CDN 파일 목록 업데이트를 위한 뮤텍스
	cdnFileListMutex sync.RWMutex
)

// 로그 메시지를 위한 구조체
type LogMessage struct {
	Level            string `json:"level"`
	Message          string `json:"message"`
	OriginURL        string `json:"origin_url"`
	Domain           string `json:"domain"`
	OriginalFilename string `json:"original_filename"`
	FilenameBase     string `json:"filename_base"`
	OriginalPathQuery string `json:"original_path_query"`
	Timestamp        string `json:"timestamp"`
}

// CDN 파일 목록을 초기화하는 함수
func initCDNFileList() {
	log.Println("[CDN File List] 초기 파일 목록 생성 중...")
	
	err := listBlobsInBucket(BucketName, 30)
	if err != nil {
		log.Printf("[CDN File List] 초기 파일 목록 생성 중 오류 발생: %v", err)
	} else {
		log.Println("[CDN File List] 초기 파일 목록 생성 완료")
	}
}

// CDN 파일 목록을 주기적으로 업데이트하는 함수
func updateCDNFileListPeriodically() {
	for {
		log.Printf("[CDN File List] 파일 목록 업데이트 시작: %s", time.Now().Format(time.RFC3339))
		
		err := listBlobsInBucket(BucketName, 0)
		if err != nil {
			log.Printf("[CDN File List] 파일 목록 업데이트 중 오류 발생: %v", err)
		} else {
			log.Printf("[CDN File List] 파일 목록 업데이트 완료: %s", time.Now().Format(time.RFC3339))
		}
		
		// 지정된 시간(1시간) 동안 대기
		time.Sleep(UpdateInterval * time.Second)
	}
}

// GCS 버킷의 파일 목록을 가져와 파일에 저장하는 함수
func listBlobsInBucket(bucketName string, timeout int) error {
	// 타임아웃 설정
	var ctx context.Context
	var cancel context.CancelFunc
	
	if timeout > 0 {
		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
		defer cancel()
	} else {
		ctx = context.Background()
	}
	
	// GCS 클라이언트 생성
	client, err := storage.NewClient(ctx, option.WithCredentialsFile("woven-province-411903-b1b12d94b3ac.json"))
	if err != nil {
		return fmt.Errorf("storage.NewClient 생성 오류: %v", err)
	}
	defer client.Close()
	
	// 버킷 접근
	bucket := client.Bucket(bucketName)
	
	// 파일 목록을 가져와 임시 파일에 저장
	tempFilePath := cdnFileListPath + ".tmp"
	tempFile, err := os.Create(tempFilePath)
	if err != nil {
		return fmt.Errorf("임시 파일 생성 오류: %v", err)
	}
	
	// 버킷의 모든 객체 목록 조회
	it := bucket.Objects(ctx, nil)
	count := 0
	
	for {
		attrs, err := it.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			tempFile.Close()
			return fmt.Errorf("객체 목록 조회 중 오류: %v", err)
		}
		
		// 파일 경로 저장
		_, err = tempFile.WriteString(attrs.Name + "\n")
		if err != nil {
			tempFile.Close()
			return fmt.Errorf("파일 쓰기 오류: %v", err)
		}
		
		count++
	}
	
	tempFile.Close()
	
	// 파일 목록 업데이트를 위한 락 획득
	cdnFileListMutex.Lock()
	defer cdnFileListMutex.Unlock()
	
	// 임시 파일을 실제 파일로 이동
	err = os.Rename(tempFilePath, cdnFileListPath)
	if err != nil {
		return fmt.Errorf("파일 이름 변경 오류: %v", err)
	}
	
	log.Printf("[CDN File List] %d개의 파일 목록을 저장했습니다", count)
	return nil
}

// 파일에서 특정 URL이 존재하는지 확인하는 함수
func searchPath(filePath, targetURL string) bool {
	cdnFileListMutex.RLock()
	defer cdnFileListMutex.RUnlock()
	
	file, err := os.Open(filePath)
	if err != nil {
		log.Printf("파일을 찾을 수 없습니다: %s, 오류: %v", filePath, err)
		return false
	}
	defer file.Close()
	
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == targetURL {
			return true
		}
	}
	
	if err := scanner.Err(); err != nil {
		log.Printf("파일 읽기 중 오류가 발생했습니다: %v", err)
	}
	
	return false
}

// 원격 서버로 로그 전송 함수
func sendLogToRemote(level, message, originalPathFull, domain, originalFilename, filenameBase, originalPathQuery string) {
	if !EnableRemoteLogging {
		return
	}
	
	logMessage := LogMessage{
		Level:            level,
		Message:          message,
		OriginURL:        originalPathFull,
		Domain:           domain,
		OriginalFilename: originalFilename,
		FilenameBase:     filenameBase,
		OriginalPathQuery: originalPathQuery,
		Timestamp:        time.Now().Format(time.RFC3339),
	}
	
	jsonData, err := json.Marshal(logMessage)
	if err != nil {
		log.Printf("[Remote Logging Error] JSON 마샬링 오류: %v", err)
		return
	}
	
	client := &http.Client{
		Timeout: time.Duration(RemoteLogTimeoutSeconds) * time.Second,
		Transport: &http.Transport{
			Proxy: nil, // 프록시 사용 안 함
		},
	}
	
	req, err := http.NewRequest("POST", RemoteLogServerURL, bytes.NewBuffer(jsonData))
	if err != nil {
		log.Printf("[Remote Logging Error] 요청 생성 오류: %v", err)
		return
	}
	
	req.Header.Set("Content-Type", "application/json")
	
	resp, err := client.Do(req)
	if err != nil {
		log.Printf("[Remote Logging Error] 로그 전송 실패 (%s): %v", RemoteLogServerURL, err)
		return
	}
	defer resp.Body.Close()
	
	if resp.StatusCode != http.StatusOK {
		log.Printf("[Remote Logging Error] 로그 전송 실패. 서버 응답: %d", resp.StatusCode)
	}
}

// 이미지 요청을 처리하는 미들웨어
type CheckAndLogMissingWebp struct{}

// ModifyRequest는 미트프록시의 요청 수정 인터페이스를 구현합니다
func (m *CheckAndLogMissingWebp) ModifyRequest(req *http.Request) error {
	// 이미지 확장자를 가진 URL인지 확인
	if originalImageExtRegex.MatchString(req.URL.Path) {
		domain := req.Host
		originalPathFull := req.URL.String()
		parsedURL, _ := url.Parse(originalPathFull)
		originalPathNoQuery := parsedURL.Path
		originalPathQuery := parsedURL.RawQuery
		originalFilename := filepath.Base(originalPathNoQuery)
		filenameBase := strings.TrimSuffix(originalFilename, filepath.Ext(originalFilename))
		webpFilename := originalPathQuery + filenameBase + ".webp"
		webpPath := fmt.Sprintf("%s/%s", domain, webpFilename)
		
		var cdnWebpURL string
		if strings.HasSuffix(CDNBaseURL, "/") {
			cdnWebpURL = CDNBaseURL + domain + "/" + webpFilename
		} else {
			cdnWebpURL = CDNBaseURL + "/" + domain + "/" + webpFilename
		}
		
		try {
			if searchPath(cdnFileListPath, webpPath) {
				fmt.Printf("\n\n%s is in list!\n\n", cdnWebpURL)
				
				// 응답을 만들어 리다이렉트
				resp := proxyutil.NewResponse(302, nil, req)
				resp.Header.Set("Location", cdnWebpURL)
				resp.Header.Set("Content-Type", "text/plain")
				resp.Header.Set("Cache-Control", "no-cache, no-store, must-revalidate")
				resp.Header.Set("Pragma", "no-cache")
				resp.Header.Set("Expires", "0")
				
				ctx := martian.NewContext(req)
				ctx.SkipRoundTrip()
				ctx.SetResponse(resp)
			} else {
				logMessage := fmt.Sprintf("[Missing WEBP] Not found at %s (Original: %s)", cdnWebpURL, originalPathFull)
				log.Print(logMessage)
				sendLogToRemote("WARN", logMessage, originalPathFull, domain, originalFilename, filenameBase, originalPathQuery)
			}
		} catch (err error) {
			logMessage := fmt.Sprintf("catch error %v (Original: %s) (cdn_webp_url: %s)", err, originalPathFull, cdnWebpURL)
			log.Print(logMessage)
			sendLogToRemote("ERROR", logMessage, originalPathFull, domain, originalFilename, filenameBase, originalPathQuery)
		}
	}
	
	return nil
}

func main() {
	// 환경 변수 설정
	os.Setenv("REQUESTS_CA_BUNDLE", "mitmproxy-ca-cert.pem")
	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "woven-province-411903-b1b12d94b3ac.json")
	
	// 초기화 및 주기적 업데이트를 위한 고루틴 시작
	go initCDNFileList()
	go updateCDNFileListPeriodically()
	
	// 프록시 생성
	p := martian.NewProxy()
	
	// TLS 설정
	tls, err := mitm.NewConfig("mitmproxy-ca-cert.pem", "mitmproxy-ca-key.pem")
	if err != nil {
		log.Fatalf("mitm.NewConfig 오류: %v", err)
	}
	
	p.SetMITM(tls)
	
	// 미들웨어 추가
	fg := fifo.NewGroup()
	fg.AddRequestModifier(&CheckAndLogMissingWebp{})
	p.SetRequestModifier(fg)
	
	// 프록시 서버 시작
	l, err := net.Listen("tcp", ":8080")
	if err != nil {
		log.Fatalf("net.Listen 오류: %v", err)
	}
	
	log.Println("프록시 서버가 8080 포트에서 실행 중입니다...")
	log.Fatal(http.Serve(l, p))
}

File: src/client/img_intercept_storelog.py
# -*- coding: utf-8 -*-
from mitmproxy import http
from mitmproxy import ctx
import re
import os
from urllib.parse import urlparse
import requests
import json
import datetime
import threading
import time
from file_exists import list_blobs_in_bucket
from google.cloud import storage
from google.oauth2 import service_account

# os.environ["REQUESTS_CA_BUNDLE"] = "mitmproxy-ca-cert.pem"
# os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "woven-province-411903-b1b12d94b3ac.json"
# Change from absolute paths to relative paths
os.environ["REQUESTS_CA_BUNDLE"] = os.path.join(os.path.dirname(os.path.abspath(__file__)), "mitmproxy-ca-cert.pem")
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = os.path.join(os.path.dirname(os.path.abspath(__file__)), "woven-province-411903-b1b12d94b3ac.json")

# 서비스 계정 키 파일 경로
KEY_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "woven-province-411903-b1b12d94b3ac.json")

os.environ["HTTP_PROXY"] = "http://127.0.0.1:8227"
os.environ["HTTPS_PROXY"] = "http://127.0.0.1:8227"


# --- User Setting ---
CDN_BASE_URL = "https://storage.cloud.google.com/cdn.ecarbon.kr" # Actual CDN address

# --- Remote Logging Setting ---
ENABLE_REMOTE_LOGGING = True # Set to True to enable remote logging
# Important: Change the IP address below to the actual IP address of PC B!
REMOTE_LOG_SERVER_URL = "http://211.253.31.134:5000/log"
REMOTE_LOG_TIMEOUT_SECONDS = 0.1 # Remote logging request timeout
# --- Remote Logging Setting End ---

# --- User Setting End ---
ORIGINAL_IMAGE_EXT_REGEX = re.compile(r"\.(png|jpe?g)(\?.*)?$|[_=](png|jpe?g)($|\?|&)|atchFileId=.*_(png|jpe?g)($|\?|&)", re.IGNORECASE)

# GCS bucket name
BUCKET_NAME = "cdn.ecarbon.kr"
# File list update interval (seconds)
UPDATE_INTERVAL = 1800  # 30 minutes = 1800 seconds
# Smaller images list file name
SMALLER_IMAGES_FILENAME = "smaller_original_images.txt"
# Smaller images list local file path
SMALLER_IMAGES_LOCAL_PATH = "smaller_original_images.txt"

## Exception handling list, will be updated...
BYPASS_DOMAINS = [
    "donga.ac.kr",
    "bufs.ac.kr",
    "www.ahnlab.com"
]

# CDN file list update function
def update_cdn_file_list_periodically():
    """Update GCS bucket file list every 30 minutes."""
    while True:
        try:
            ctx.log.info(f"[CDN File List] CDN file list update started: {datetime.datetime.now().isoformat()}")
            list_blobs_in_bucket(BUCKET_NAME)
            ctx.log.info(f"[CDN File List] CDN file list update completed: {datetime.datetime.now().isoformat()}")
        except Exception as e:
            ctx.log.error(f"[CDN File List] CDN file list update error: {e}")
        
        # Wait for the specified time (30 minutes)
        time.sleep(UPDATE_INTERVAL)

# Download smaller_original_images.txt file from GCS bucket
def download_smaller_images_list(bucket_name, timeout=None):
    """Download smaller_original_images.txt file from GCS bucket."""
    try:
        # 서비스 계정 자격 증명 명시적 로드
        credentials = service_account.Credentials.from_service_account_file(KEY_PATH)
        
        # Initialize client with explicit credentials
        storage_client = storage.Client(credentials=credentials)
        # set timeout
        storage_client._http.timeout = timeout if timeout else None
        
        bucket = storage_client.bucket(bucket_name)
        blob = bucket.blob(SMALLER_IMAGES_FILENAME)
        
        # Check if the file exists
        if blob.exists():
            # Download the file
            blob.download_to_filename(SMALLER_IMAGES_LOCAL_PATH)
            ctx.log.info(f"[Smaller Images] {SMALLER_IMAGES_FILENAME} file downloaded successfully")
            return True
        else:
            ctx.log.warn(f"[Smaller Images] {SMALLER_IMAGES_FILENAME} file not found in bucket")
            # Create an empty file locally
            with open(SMALLER_IMAGES_LOCAL_PATH, 'w', encoding='utf-8') as f:
                f.write("domain,original_url,recorded_at\n")
            return False
    except Exception as e:
        ctx.log.error(f"[Smaller Images] smaller_original_images.txt download error: {e}")
        return False

# Update smaller_original_images.txt file periodically
def update_smaller_images_list_periodically():
    """Update smaller_original_images.txt file periodically."""
    while True:
        try:
            ctx.log.info(f"[Smaller Images] smaller_original_images.txt update started: {datetime.datetime.now().isoformat()}")
            download_smaller_images_list(BUCKET_NAME)
            ctx.log.info(f"[Smaller Images] smaller_original_images.txt update completed: {datetime.datetime.now().isoformat()}")
        except Exception as e:
            ctx.log.error(f"[Smaller Images] smaller_original_images.txt update error: {e}")
        
        # Wait for the specified time
        time.sleep(UPDATE_INTERVAL)

# Program startup initial file list update - run in the background thread to prevent blocking
def init_cdn_file_list():
    ctx.log.info("[CDN File List] Initial CDN file list creation...")
    try:
        list_blobs_in_bucket(BUCKET_NAME, timeout=15)
        ctx.log.info("[CDN File List] Initial CDN file list created")
        
        # Download smaller_original_images.txt file from GCS bucket
        download_smaller_images_list(BUCKET_NAME, timeout=15)
    except Exception as e:
        ctx.log.error(f"[CDN File List] initial CDN file list creation error: {e}")

# Initialize CDN file list at program start - run in the background thread to prevent blocking
init_thread = threading.Thread(target=init_cdn_file_list, daemon=True)
init_thread.start()
ctx.log.info("[CDN File List] initial CDN file list creation thread started")

# Periodic CDN file list update thread
updater_thread = threading.Thread(target=update_cdn_file_list_periodically, daemon=True)
updater_thread.start()
ctx.log.info("[CDN File List] periodic CDN file list update thread started")

# Periodic smaller_original_images.txt update thread
smaller_images_thread = threading.Thread(target=update_smaller_images_list_periodically, daemon=True)
smaller_images_thread.start()
ctx.log.info("[Smaller Images] periodic smaller_original_images.txt update thread started")

# Find URL in file
def search_path(file_path, Target_URL):
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            for line_number, line in enumerate(file, 1):
                line = line.strip()  # Remove trailing whitespace and newline characters
                if Target_URL in line:
                    return True
            # All lines checked but not found
            return False
    except FileNotFoundError:
        print(f"File not found: {file_path}")
        return False
    except Exception as e:
        print(f"Error reading file: {e}")
        return False

# Check if the original image URL is in 'smaller_original_images.txt'
def is_smaller_original_image(original_url):
    """Check if the original image URL is in 'smaller_original_images.txt'"""
    try:
        with open(SMALLER_IMAGES_LOCAL_PATH, 'r', encoding='utf-8') as file:
            # Skip the first line (header)
            next(file, None)
            
            for line in file:
                if line.strip() and original_url in line:
                    return True
        return False
    except FileNotFoundError:
        ctx.log.warn(f"[Smaller Images] File not found: {SMALLER_IMAGES_LOCAL_PATH}")
        return False
    except Exception as e:
        ctx.log.error(f"[Smaller Images] Error reading file: {e}")
        return False




# --- Remote logging function ---
def send_log_to_remote(level, message, original_path_full, domain, original_filename, filename_base, original_path_query):
    """Send log message to specified server using POST request."""
    if not ENABLE_REMOTE_LOGGING:
        return
    try:
        payload = {
            "level": level,
            "message": message,
            "origin_url": original_path_full,
            "domain": domain,
            "original_filename": original_filename,
            "filename_base": filename_base,
            "original_path_query": original_path_query,
            "timestamp": datetime.datetime.now().isoformat() # Current time added
        }
        headers = {'Content-Type': 'application/json'}

        # Set proxy to None to avoid using proxy
        proxies = {
          "http": None,
          "https": None,
        }

        # ctx.log.info(f"Attempting to send log DIRECTLY to {REMOTE_LOG_SERVER_URL}") # 디버깅 로그 추가 (선택사항)

        response = requests.post(
            REMOTE_LOG_SERVER_URL,
            headers=headers,
            data=json.dumps(payload),
            timeout=REMOTE_LOG_TIMEOUT_SECONDS,
            proxies=proxies
        )

        if response.status_code != 200:
            ctx.log.error(f"[Remote Logging Error] Failed to send log. Server responded with {response.status_code}")
    except requests.exceptions.RequestException as e:
        # Improve error log: include proxy information in the full error output
        ctx.log.error(f"[Remote Logging Error] Failed to send log to {REMOTE_LOG_SERVER_URL}: {e}")
    except Exception as e:
        ctx.log.error(f"[Remote Logging Error] Unexpected error during remote logging: {e}")



class CheckAndLogMissingWebp:
    ctx.options.ignore_hosts = BYPASS_DOMAINS
    def request(self, flow: http.HTTPFlow) -> None:    
        match = ORIGINAL_IMAGE_EXT_REGEX.search(flow.request.path)
        if match:
            domain = flow.request.pretty_host
            original_path_full = flow.request.path # Original URL
            parsed_url = urlparse(original_path_full)
            original_path_no_query = parsed_url.path
            original_path_query = parsed_url.query
            original_filename = os.path.basename(original_path_no_query)
            filename_base, _ = os.path.splitext(original_filename)
            webp_filename = original_path_query + filename_base + ".webp"
            webp_path = f"{domain}/{webp_filename}"

            # Path rule : CDN_BASE_URL / domain / original path query + original filename.webp
            if CDN_BASE_URL.endswith('/'):
                cdn_webp_url = CDN_BASE_URL + domain + '/' + webp_filename
            else:
                cdn_webp_url = CDN_BASE_URL + '/' + domain + '/' + webp_filename

            try:
                # Check if the original image is in 'smaller_original_images.txt'
                if is_smaller_original_image(original_path_full):
                    print(f"[Smaller Original] Original image found in smaller_original_images.txt: {original_path_full}")
                    # # If the original image is in 'smaller_original_images.txt', use the original image
                    # flow.response = http.Response.make(
                    #     302,  # Status code: Found (Temporary Redirect)
                    #     b"",  # Response body
                    #     {
                    #     "Location": original_path_full, # Use the original image URL
                    #     "Content-Type": "text/plain",
                    #     "Cache-Control": "public, max-age=3600",
                    #     }
                    # )
                elif search_path("cdn_file_list.txt", webp_path):
                    print(f"\n\n{cdn_webp_url} is in list!\n\n")
                    flow.response = http.Response.make(
                        302,  # Status code: Found (Temporary Redirect)
                        b"",  # Response body
                        {
                        "Location": cdn_webp_url, # Final CDN URL (only .webp filename)
                        "Content-Type": "text/plain",
                        "Cache-Control": "public, max-age=3600",
                        # "Pragma": "no-cache",
                        # "Expires": "0",
                        }
                    )
                    log_message = (
                        f"[Found WEBP] Found at {cdn_webp_url} "
                        f"(Original: {original_path_full})"
                    )
                    send_log_to_remote("SUCCESS", log_message, original_path_full, domain, original_filename, filename_base, original_path_query)
                else:
                    log_message = (
                        f"[Missing WEBP] Not found at {cdn_webp_url} "
                        f"(Original: {original_path_full})"
                    )
                    # Local log output
                    ctx.log.warn(log_message)
                    # Send log to remote server
                    send_log_to_remote("WARN", log_message, original_path_full, domain, original_filename, filename_base, original_path_query)
            except Exception as e:
                log_message = (
                    f"catch error {e}"
                    f"(Original: {original_path_full})"
                    f"(cdn_webp_url: {cdn_webp_url})"
                )
                # Local log output
                ctx.log.error(log_message)
                # Send log to remote server
                send_log_to_remote("ERROR", log_message, original_path_full, domain, original_filename, filename_base, original_path_query)

addons = [
    CheckAndLogMissingWebp()
]